src/geohash_helper.c

```diff
@@ -100,34 +100,41 @@ uint8_t geohashEstimateStepsByRadius(double range_meters, double lat) {
  * Since this function is currently only used as an optimization, the
  * optimization is not used for very big radiuses, however the function
  * should be fixed. */
-int geohashBoundingBox(double longitude, double latitude, double radius_meters,
-                       double *bounds) {
+int geohashBoundingBox(GeoShape *shape, double *bounds) {
     if (!bounds) return 0;
-
-    bounds[0] = longitude - rad_deg(radius_meters/EARTH_RADIUS_IN_METERS/cos(deg_rad(latitude)));
-    bounds[2] = longitude + rad_deg(radius_meters/EARTH_RADIUS_IN_METERS/cos(deg_rad(latitude)));
-    bounds[1] = latitude - rad_deg(radius_meters/EARTH_RADIUS_IN_METERS);
-    bounds[3] = latitude + rad_deg(radius_meters/EARTH_RADIUS_IN_METERS);
+    double longitude = shape->xy[0];
+    double latitude = shape->xy[1];
+    double height = shape->conversion * (shape->type == CIRCULAR_TYPE ? shape->radius : shape->height);
+    double width = shape->conversion * (shape->type == CIRCULAR_TYPE ? shape->radius : shape->width);
+
+    bounds[0] = longitude - rad_deg(height/EARTH_RADIUS_IN_METERS/cos(deg_rad(latitude)));
+    bounds[2] = longitude + rad_deg(height/EARTH_RADIUS_IN_METERS/cos(deg_rad(latitude)));
+    bounds[1] = latitude - rad_deg(width/EARTH_RADIUS_IN_METERS);
+    bounds[3] = latitude + rad_deg(width/EARTH_RADIUS_IN_METERS);
     return 1;
 }
 
 /* Return a set of areas (center + 8) that are able to cover a range query
  * for the specified position and radius. */
-GeoHashRadius geohashGetAreasByRadius(double longitude, double latitude, double radius_meters) {
+GeoHashRadius geohashGetAreasByRadius(GeoShape *shape) {
     GeoHashRange long_range, lat_range;
     GeoHashRadius radius;
     GeoHashBits hash;
     GeoHashNeighbors neighbors;
     GeoHashArea area;
     double min_lon, max_lon, min_lat, max_lat;
-    double bounds[4];
     int steps;
 
-    geohashBoundingBox(longitude, latitude, radius_meters, bounds);
-    min_lon = bounds[0];
-    min_lat = bounds[1];
-    max_lon = bounds[2];
-    max_lat = bounds[3];
+    geohashBoundingBox(shape, shape->bounds);
+    min_lon = shape->bounds[0];
+    min_lat = shape->bounds[1];
+    max_lon = shape->bounds[2];
+    max_lat = shape->bounds[3];
+
+    double longitude = shape->xy[0];
+    double latitude = shape->xy[1];
+    double radius_meters = shape->conversion * (shape->type == CIRCULAR_TYPE ? shape->radius :
```

Comment:
given that `geohashEstimateStepsByRadius` will return an higher number of steps for small distances, isn't it enough to use as input for `RECTANGLE_TYPE` the minimum of width and height?

Reply:
I think the longest radius in RECTANGLE (or other shapes) should be chosen.

For example, width=200km, height=1km, if steps are calculated according to height, the black area will miss the search.

![image](https://user-images.githubusercontent.com/13137470/100299404-908f2200-2fce-11eb-8ba6-a7c6bdf758d9.png)

Reply:
@yangbodong22011 if we look at the `geohashEstimateStepsByRadius()` code we see that the smaller the difference the larger the steps ( meaning precision ). If i understood correctly this function is not changing ranges of search but only increasing/decreasing precision based on distance. If we agree on that, then using the smallest distance should be enough. wdyt? 
```
/* This function is used in order to estimate the step (bits precision)
 * of the 9 search area boxes during radius queries. */
uint8_t geohashEstimateStepsByRadius(double range_meters, double lat) {
    if (range_meters == 0) return 26;
    int step = 1;
    while (range_meters < MERCATOR_MAX) {
        range_meters *= 2;
        step++;
    }
```

Reply:
@filipecosta90 Let's focused on a case:

After compiling with the current code, execute the following command, can find the result
```
127.0.0.1:6379> GEOADD Sicily 12.75 37.00 "test"
(integer) 1
127.0.0.1:6379> GEOBBOX Sicily 15 37 1 200 km WITHDIST
1) 1) "test"
   2) "199.8611"
127.0.0.1:6379>
```


Next modify the code, compile and restart redis

```
diff --git a/src/geohash_helper.c b/src/geohash_helper.c
index 297b1e362..59e42a4f3 100644
--- a/src/geohash_helper.c
+++ b/src/geohash_helper.c
@@ -136,7 +136,7 @@ GeoHashRadius geohashGetAreasByRadius(GeoShape *shape) {
     double longitude = shape->xy[0];
     double latitude = shape->xy[1];
     double radius_meters = shape->conversion * (shape->type == CIRCULAR_TYPE? shape->radius:
- sqrt(shape->width * shape->width + shape->height * shape->height));
+ shape->width> shape->height? shape->height: shape->width);
 

```

Execute the same query again:

```
127.0.0.1:6379> GEOBBOX Sicily 15 37 1 200 km WITHDIST
(empty list or set)
127.0.0.1:6379>
```
Will not find test